pkg/queue/concurrency_state_test.go

```diff
@@ -17,116 +17,127 @@ limitations under the License.
+func TestConcurrencyStateHandlerParallelSubsumed(t *testing.T) {
+	paused := atomic.NewInt64(0)
+	resumed := atomic.NewInt64(0)
 
-			if got, want := paused.Load(), tt.pauses; got != want {
-				t.Errorf("expected to be paused %d times, but was paused %d times", want, got)
-			}
+	req1 := make(chan struct{})
+	handler := func(w http.ResponseWriter, r *http.Request) {
+		if r.URL.Query().Get("req") == "1" {
+			req1 <- struct{}{} // to know it's here.
+			req1 <- struct{}{} // to make it wait.
+		}
+	}
+	logger := ltesting.TestLogger(t)
+	h := ConcurrencyStateHandler(logger, http.HandlerFunc(handler), func() { paused.Inc() }, func() { resumed.Inc() })
 
-			if got, want := delegated.Load(), int64(len(tt.events)); got != want {
-				t.Errorf("expected to be delegated %d times, but delegated %d times", want, got)
-			}
+	go func() {
+		defer func() { req1 <- struct{}{} }()
+		h.ServeHTTP(httptest.NewRecorder(), httptest.NewRequest("GET", fmt.Sprintf("http://target?req=%d", 1), nil))
+	}()
 
-			if got, want := resumed.Load(), tt.resumes; got != want {
-				t.Errorf("expected to be resumed %d times, but was resumed %d times", want, got)
-			}
-		})
+	<-req1 // Wait for req1 to arrive.
+
+	// Send a second request, which can pass through.
+	h.ServeHTTP(httptest.NewRecorder(), httptest.NewRequest("GET", "http://target", nil))
+
+	<-req1 // Allow req1 to pass.
+	<-req1 // Wait for req1 to finish.
+
+	if got, want := pollFor(paused, 1), int64(1); got != want {
+		t.Errorf("Pause was called %d times, want %d times", got, want)
+	}
+
+	if got, want := pollFor(resumed, 1), int64(1); got != want {
+		t.Errorf("Resume was called %d times, want %d times", got, want)
+	}
+}
+
+func TestConcurrencyStateHandlerParallelOverlapping(t *testing.T) {
+	paused := atomic.NewInt64(0)
+	resumed := atomic.NewInt64(0)
+
+	req1 := make(chan struct{})
+	req2 := make(chan struct{})
+	handler := func(w http.ResponseWriter, r *http.Request) {
+		if r.URL.Query().Get("req") == "1" {
+			req1 <- struct{}{} // to know it's here.
+			req1 <- struct{}{} // to make it wait.
+		} else {
+			req2 <- struct{}{} // to know it's here.
+			req2 <- struct{}{} // to make it wait.
+		}
+	}
+	logger := ltesting.TestLogger(t)
+	h := ConcurrencyStateHandler(logger, http.HandlerFunc(handler), func() { paused.Inc() }, func() { resumed.Inc() })
+
+	go func() {
+		defer func() { req1 <- struct{}{} }()
+		h.ServeHTTP(httptest.NewRecorder(), httptest.NewRequest("GET", fmt.Sprintf("http://target?req=%d", 1), nil))
+	}()
+
+	<-req1 // Wait for req1 to arrive.
+
+	go func() {
+		defer func() { req2 <- struct{}{} }()
+		h.ServeHTTP(httptest.NewRecorder(), httptest.NewRequest("GET", fmt.Sprintf("http://target?req=%d", 2), nil))
+	}()
+
+	<-req2 // Wait for req2 to arrive
+
+	<-req1 // Allow req1 to pass.
+	<-req1 // Wait for req1 to finish.
+
+	<-req2 // Allow req2 to pass.
+	<-req2 // Wait for req2 to finish.
+
+	if got, want := pollFor(paused, 1), int64(1); got != want {
```

Comment:
since pollFor exits as soon as it sees the expected value, how do we now we _wouldnt_ see 2 here, if we waited for longer?

Reply:
Technically, we don't.
FWIW, the speed-improvement PR will make this obsolete as the state will be consistent with the ServeHTTP calls there.

Reply:
Okiedoke, stamping